drivers/adc/adc_shell.c

```diff
@@ -0,0 +1,447 @@
+/*
+ * Copyright (c) 2018 Prevas A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <shell/shell.h>
+#include <stdlib.h>
+#include <adc.h>
+#include <ctype.h>
+#include <misc/util.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_shell);
+
+struct adc_hdl {
+	char *device_name;
+	struct adc_channel_cfg channel_config;
+	u8_t resolution;
+};
+
+struct adc_hdl adc_list[] = {
+#ifdef CONFIG_ADC_0
+	{
+		.device_name = DT_ADC_0_NAME,
+		.channel_config = {
+			.gain = ADC_GAIN_1,
+			.reference = ADC_REF_INTERNAL,
+			.acquisition_time = ADC_ACQ_TIME_DEFAULT,
+			.channel_id = 0,
+		},
+		.resolution = 0,
+	},
+#endif
+#ifdef CONFIG_ADC_1
+	{
+		.device_name = DT_ADC_1_NAME,
+		.channel_config = {
+			.gain = ADC_GAIN_1,
+			.reference = ADC_REF_INTERNAL,
+			.acquisition_time = ADC_ACQ_TIME_DEFAULT,
+			.channel_id = 0,
+		},
+		.resolution = 0,
+	},
+#endif
+};
+
+struct args_index {
+	s8_t adc;
+	s8_t parent_adc;
+	u8_t channel;
+	u8_t conf;
+	u8_t acq_unit;
+};
+
+struct args_number {
+	u8_t help;
+	u8_t channel;
+	u8_t acq_time;
+	u8_t resolution;
+	u8_t read;
+};
+
+static const struct args_index args_indx = {
+	.adc = -1,
+	.parent_adc = -2,
+	.channel = 1,
+	.conf = 1,
+	.acq_unit = 2,
+};
+
+static const struct args_number args_no = {
+	.help = 1,
+	.channel = 2,
+	.acq_time = 3,
+	.resolution = 2,
+	.read = 2,
+};
+
+/** get_adc_from_list returns the number entry of the adc in the adc_list,
+ * returns -ENODEV if it doesn't exist
+ */
+static int get_adc_from_list(char *name)
+{
+	int adc_idx;
+
+	for (adc_idx = 0; adc_idx < ARRAY_SIZE(adc_list); adc_idx++) {
+		if (!strcmp(name, adc_list[adc_idx].device_name)) {
+			return adc_idx;
+		}
+	}
+	return -ENODEV;
+}
+
+static int cmd_adc_channel(const struct shell *shell, size_t argc, char **argv)
+{
+	int retval = 0;
+	struct device *adc_dev;
+	int chosen_adc;
+
+	if (argc != args_no.channel) {
+		shell_fprintf(shell, SHELL_NORMAL,
+				"Usage: channel <channel_id>\n");
+		return 0;
+	}
+
+	chosen_adc = get_adc_from_list(argv[args_indx.adc]);
+	if (chosen_adc < 0) {
+		return -EINVAL;
```

Comment:
Please add a `shell_error()` here as well. It's a good idea to always notify the caller of a shell command when a command fails.